sui_core/src/authority/authority_checkpoints.rs

```diff
@@ -0,0 +1,1325 @@
+
+        // Does the fragment event suggest it is for the current round?
+        let next_checkpoint_seq = self.next_checkpoint();
+        fp_ensure!(
+            *_fragment.proposer.0.checkpoint.sequence_number() == next_checkpoint_seq,
+            SuiError::GenericAuthorityError {
+                error: format!(
+                    "Incorrect sequence number, expected {}",
+                    next_checkpoint_seq
+                )
+            }
+        );
+
+        // Only a fragment that involves ourselves to be sequenced through
+        // this node.
+        fp_ensure!(
+            _fragment.proposer.0.authority == self.name || _fragment.other.0.authority == self.name,
+            SuiError::GenericAuthorityError {
+                error: "Fragment does not involve this node".to_string(),
+            }
+        );
+
+        // Save in the list of local fragments for this sequence.
+        let other_name = if _fragment.proposer.0.authority == self.name {
+            _fragment.other.0.authority
+        } else {
+            _fragment.proposer.0.authority
+        };
+        if !self.local_fragments.contains_key(&other_name)? {
+            self.local_fragments.insert(&other_name, _fragment)?;
+        } else {
+            // We already have this fragment, so we can ignore it.
+            return Err(SuiError::GenericAuthorityError {
+                error: format!("Already processed fragment with {:?}", other_name),
+            });
+        }
+
+        // TODO: Checks here that the fragment makes progress over the existing
+        //       construction of components using the self.fragments table. This
+        //       is an optimization for later.
+
+        let locals = self.get_locals();
+        if !locals.no_more_fragments {
+            // Send to consensus for sequencing.
+            if let Some(sender) = &self.sender {
+                sender.send_to_consensus(_fragment.clone())?;
+            } else {
+                return Err(SuiError::GenericAuthorityError {
+                    error: "No consensus sender configured".to_string(),
+                });
+            }
+        } else {
+            // Maybe the fragment we received allows us to complete the current checkpoint?
+            // Since we seem to be missing information to complete it (ie there is a checkpoint
+            // but we are not inlcuded in it.)
+            loop {
+                let construct = self.attempt_to_construct_checkpoint();
+                // Exit if checkpoint construction leads to an error or returns false
+                // (ie no new checkpoint is created.)
+                if construct.is_err() || !construct.unwrap() {
+                    break;
+                }
+            }
+        }
+
+        // NOTE: we should charge the node that sends this into consensus
+        //       according to the byte length of the fragment, to create
+        //       incentives for nodes to submit smaller fragments.
+
```

Comment:
Is it possible that validators may be incentivized to keep as much state as possible out of checkpointing then? Submitting a checkpoint that's entirely valid but makes no forward progress seems to be a honest behavior that results in less cost.

Reply:
Based on hour invariants not submitting a signed-effect certificate to a checkpoint is a slashable offense among the set of validators whose fragments are in the checkpoint and the certificate signers. It actually violates safety of single-owner transactions.

Reply:
Why would this violate safety of single-owner objects? I though only liveness would be violated (what am I missing?)